### PR TITLE
Disable ability to provide relative paths in sources

### PR DESCRIPTION
**Pull request by BenWilson2**
#8273

## What changes are proposed in this pull request?

Disable the ability to provide relative paths in the source field when creating model versions.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/server/handlers.py [hunk 1]
```diff
@@ -1352,6 +1378,9 @@ def _validate_source(source: str, run_id: str) -> None:
             INVALID_PARAMETER_VALUE,
         )
 
+    if is_http_uri(source) or is_mlflow_artifacts_uri(source) or is_local_uri(source):
+        _validate_non_local_source_contains_relative_paths(source)
```
Comment (harupy) [inline mlflow/server/handlers.py:1382]:
```suggestion
    _validate_non_local_source_contains_relative_paths(source)
```

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/8281/files#r1171980058 is addressed :)

File: mlflow/server/handlers.py (hunk 1 above)
Comment (BenWilson2) [reply mlflow/server/handlers.py]: Great idea :) Updated, changed the Exception message, and added a few more conditions in the test suite based on our discussion (ftp, s3, etc)

File: mlflow/server/handlers.py [hunk 2]
```diff
@@ -1323,6 +1323,34 @@ def _delete_registered_model_tag():
...
+    "https://host:port/api/2.0/mlflow-artifacts/artifacts/../../../../"
+    "/models/artifacts/../../../"
+    "s3:/my_bucket/models/path/../../other/path"
+    "file://path/to/../../../../some/where/you/should/not/be"
+    """
+    source_path = urllib.parse.urlparse(source).path
+    raw_source_path = pathlib.Path(source_path)
+    resolved_path = raw_source_path.resolve()
+
+    if resolved_path != raw_source_path:
```
Comment (harupy) [inline mlflow/server/handlers.py:1345]: It looks like this doesn't work on Windows...

Comment (BenWilson2) [reply mlflow/server/handlers.py]: looks like `WindowsPath().resolve()` prepends the pwd drive designator. Updated the logic and validated it works on my Windows PC.

**Pull request merged**